### PR TITLE
simplify the remove_exceptions API

### PR DESCRIPTION
**Pull request by kroening**
This removes the usage of a pointer an the potential to call the variant
that does remove_instanceof without class hierarchy.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: jbmc/src/java_bytecode/remove_exceptions.cpp [hunk 1]
```diff
@@ -87,14 +87,14 @@ class remove_exceptionst
 
   explicit remove_exceptionst(
     symbol_table_baset &_symbol_table,
-    const class_hierarchyt *class_hierarchy,
```
Comment (smowton) [inline jbmc/src/java_bytecode/remove_exceptions.cpp:90]: These changes are unnecessary, the init list knows the difference between locals and members

File: jbmc/src/java_bytecode/remove_exceptions.cpp [hunk 2]
```diff
@@ -593,24 +593,85 @@ void remove_exceptionst::operator()(goto_programt &goto_program)
...
+void remove_exceptions(
+  goto_modelt &goto_model,
+  message_handlert &message_handler)
+{
+  remove_exceptions(
+    goto_model.symbol_table, goto_model.goto_functions, message_handler);
+}
+
+/// removes throws/CATCH-POP/CATCH-PUSH
+void remove_exceptions_and_instanceof(
```
Comment (smowton) [inline jbmc/src/java_bytecode/remove_exceptions.cpp]: Name not really accurate; this should only affect the newly introduced instanceof instructions, not existing ones

File: jbmc/src/java_bytecode/remove_exceptions.h [hunk 3]
```diff
@@ -23,28 +23,34 @@ Date:   December 2016
...
+/// Any instanceof expressions are expanded.
+void remove_exceptions_and_instanceof(
+  goto_programt &,
+  symbol_table_baset &,
+  const class_hierarchyt &,
+  message_handlert &);
+
+/// Removes 'throw x' and CATCH-PUSH/CATCH-POP
+/// and adds the required instrumentation (GOTOs and assignments)
+/// Any instanceof expressions are expanded.
```
Comment (smowton) [inline jbmc/src/java_bytecode/remove_exceptions.h]: Any *new* instanceof instructions

File: jbmc/src/java_bytecode/remove_exceptions.cpp (hunk 1 above)
Comment (kroening) [reply jbmc/src/java_bytecode/remove_exceptions.cpp:90]: Consistency, mostly.

File: jbmc/src/java_bytecode/remove_exceptions.cpp (hunk 2 above)
Comment (kroening) [reply jbmc/src/java_bytecode/remove_exceptions.cpp]: Ok, will flip naming around.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 0cdc1f039c568d392a63b37f855284993fc3b07b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90551420
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: jbmc/src/java_bytecode/remove_exceptions.cpp (hunk 1 above)
Comment (peterschrammel) [reply jbmc/src/java_bytecode/remove_exceptions.cpp:90]: Shouldn't also message_handler be renamed then?

**Review by peterschrammel — APPROVED** [review]
@thk123, @allredj, this is going to break TG, but trivial to fix.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 902cc69f71ace0109cf12928469d5a68b29893d1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90679994
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: jbmc/src/java_bytecode/remove_exceptions.cpp (hunk 1 above)
Comment (kroening) [reply jbmc/src/java_bytecode/remove_exceptions.cpp:90]: Done

**Comment by tautschnig** [comment]
Please rebase to re-trigger the bot.

**Comment by thk123** [comment]
I wouldn't bother - this needs a manual bump, I believe @allredj is looking at this.

**Comment by tautschnig** [comment]
Marked do-not-merge and assigned to @allredj.

**Comment by allredj** [comment]
Test-gen PR is here: https://github.com/diffblue/test-gen/pull/2457. Will notify when passing.

**Comment by allredj** [comment]
TG bump passing. Good to merge.

**Pull request merged**